python/repositories.bzl

```diff
@@ -1,8 +1,164 @@
+# Copyright 2022 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
 """This file contains macros to be called during WORKSPACE evaluation.
 
 For historic reasons, pip_repositories() is defined in //python:pip.bzl.
 """
 
+load("//python/private:toolchains_repo.bzl", "PLATFORMS", "toolchains_repo")
+load("//python/private:versions.bzl", "MINOR_MAPPING", "RELEASE_DATE", "RELEASE_URL", "TOOL_VERSIONS")
+
 def py_repositories():
     # buildifier: disable=print
     print("py_repositories is a no-op and is deprecated. You can remove this from your WORKSPACE file")
+
+########
+# Remaining content of the file is only used to support toolchains.
+########
+_DOC = "Fetch external tools needed for python toolchain."
+_ATTRS = {
+    "platform": attr.string(mandatory = True, values = PLATFORMS.keys()),
+    "python_version": attr.string(mandatory = True, values = TOOL_VERSIONS.keys() + MINOR_MAPPING.keys()),
+}
+
+def _python_repo_impl(rctx):
+    version = rctx.attr.python_version
+    if version in MINOR_MAPPING:
+        version = MINOR_MAPPING[version]
+    release_filename = "cpython-{version}-{platform}-pgo+lto-{release_date}.tar.zst".format(
+        release_date = RELEASE_DATE,
+        version = version,
+        platform = rctx.attr.platform,
+    )
+    url = "{release_url}/{release_filename}".format(
+        release_url = RELEASE_URL,
+        release_filename = release_filename,
+    )
+    rctx.download(
+        url = url,
+        integrity = TOOL_VERSIONS[version][rctx.attr.platform],
+        output = release_filename,
+    )
+    unzstd = rctx.which("unzstd")
```

Comment:
This is the bit that doesn't feel production-ready. We either inspect the host system for the needed binary, or download the source and compile it on the host. (Do Windows machines have `make` available by default?)

So if we don't want to adopt this non-hermeticity, what options do we have?

* Try and land non-zstd tarball distributions in the python-standalone repo. Bazel natively supports regular tarballs.
    *  Do we know if Gregory is against using it? 
* Get zstd support landed in Bazel. https://github.com/bazelbuild/bazel/pull/11968 is an old PR, but there's been activity from Googlers on it in the last 18 days (mid-Jan 2022) and it seems like it might be mergeable?

Reply:
This is no different than using `gzip` with `tar`. I don't see anyone worried about which version of `gzip` the system is using to decompress a `.tar.gz` file (at least not at this level).

I.e. if `unzstd` exits code 0, it successfully extracted the software we want - and that is deterministic. For Windows we can follow a similar approach to compiling it without Make if it's not supported on that system.

Reply:
just FYI internally we compile a zstd cc_binary that we use to extract the python toolchain http download.

So making zstd binary configurable is very much desirable.

Let me know if you are interested in the BUILD file for zstd

Reply:
https://github.com/grailbio/py_singlefile_binary/tree/main/third_party%2Fzstd
And here is how we put it together https://github.com/grailbio/py_singlefile_binary/blob/main/bazel/zstd/rules.bzl

Reply:
`zstd` decompression isn’t necessary since [this](https://github.com/indygreg/python-build-standalone/issues/79) issue was actioned, which added `.tar.gz` release artefacts.

Reply:
+1 for using tarballs

Reply:
Thanks for the heads up @jheaff1, that's awesome.

Reply:
So they latest release publishes the 'install only' tarballs that we want, for Python 3.9 and 3.10, and each of those for Windows, Apple, and Linux. 

e.g [github.com/indygreg/python-build-standalone/releases/download/20211017/cpython-3.10.0-x86_64-apple-darwin-install_only-20211017T1616.tar.gz](https://github.com/indygreg/python-build-standalone/releases/download/20211017/cpython-3.10.0-x86_64-apple-darwin-install_only-20211017T1616.tar.gz)

@f0rmiga the 'install only' tarballs will remove a decent amount of complexity and non-hermeticity. Let's give them a try.

Reply:
Just a heads up that I'm getting hands on with the 'install only' tarballs over in https://github.com/thundergolfer/example-bazel-monorepo/pull/54 and documenting my troubles.